core/src/Streamly/Internal/Data/MutArray/Type.hs

```diff
@@ -2582,7 +2596,240 @@ byteEq :: MonadIO m => MutArray a -> MutArray a -> m Bool
+    D.Stream step' (SpliceInitial state)
+
+    where
+
+    functionName = "Streamly.Internal.Data.MutArray.rCompactChunksLE"
+
+    {-# INLINE_LATE step' #-}
+    step' gst (SpliceInitial st) = do
+        when (n <= 0) $
+            -- XXX we can pass the module string from the higher level API
+            error $ functionName ++ ": the size of arrays [" ++ show n
+                ++ "] must be a natural number"
+        r <- step gst st
+        case r of
+            D.Yield arr s -> return $
+                let len = byteLength arr
+                 in if len >= n
+                    then D.Skip (SpliceYielding arr (SpliceInitial s))
+                    else D.Skip (SpliceBuffering s arr)
+            D.Skip s -> return $ D.Skip (SpliceInitial s)
+            D.Stop -> return D.Stop
+
+    step' gst (SpliceBuffering st buf) = do
+        r <- step gst st
+        case r of
+            D.Yield arr s -> do
+                let len = byteLength buf + byteLength arr
+                if len > n
+                then return $
+                    D.Skip (SpliceYielding buf (SpliceBuffering s arr))
+                else do
+                    buf1 <- splice buf arr
+                    return $ D.Skip (SpliceBuffering s buf1)
+            D.Skip s -> return $ D.Skip (SpliceBuffering s buf)
+            D.Stop -> return $ D.Skip (SpliceYielding buf SpliceFinish)
+
+    step' _ SpliceFinish = return D.Stop
+
+    step' _ (SpliceYielding arr next) = return $ D.Yield arr next
+
+-- | Fold @fCompactChunksGE n@ coalesces adjacent arrays in the input stream
+-- until the size becomes greater than or equal to n.
+--
+{-# INLINE_NORMAL fCompactChunksGE #-}
+fCompactChunksGE ::
+       forall m a. (MonadIO m, Unbox a)
+    => Int -> FL.Fold m (MutArray a) (MutArray a)
+fCompactChunksGE n = Fold step initial extract extract
+
+    where
+
+    nBytes = n * SIZE_OF(a)
+
+    functionName = "Streamly.Internal.Data.MutArray.fCompactChunksGE"
+
+    initial =
+        return
+            $ if n < 0
+              then error
+                       $ functionName
+                       ++ ": the size of arrays ["
+                       ++ show n ++ "] must be a natural number"
+              else FL.Partial Nothing
+
+    step Nothing arr =
+        return
+            $ let len = byteLength arr
+               in if len >= nBytes
+                  then FL.Done arr
+                  else FL.Partial (Just arr)
+    step (Just buf) arr = do
+        let len = byteLength buf + byteLength arr
+        buf1 <- splice buf arr
+        if len >= n
+        then return $ FL.Done buf1
+        else return $ FL.Partial (Just buf1)
+
+    extract Nothing = return nil
+    extract (Just buf) = return buf
+
+-- | Like 'compactChunksGE' but for transforming folds instead of stream.
+--
+-- >>> lCompactChunksGE n = Fold.many (MutArray.fCompactChunksGE n)
+--
+{-# INLINE_NORMAL lCompactChunksGE #-}
+lCompactChunksGE :: (MonadIO m, Unbox a)
```

Comment:
Can we move this to the fold module?

Reply:
No. this is specific to an array type (e.g. Array, Array.Generic, MutArray, MutArray.Generic etc), you cannot have all these functions in Fold module. Also Fold module should not have functions for all types, the module for MutArray type should have folds operating on this type.